### PR TITLE
Add regex flags parameter to python cudf strings split

### LINKED ISSUE #3584: [FEA] Support regular expression patterns for string split
I'd like to be able to pass a regular expression pattern for string splitting, like I could do in base Python or pandas. This is relevant for tasks like trying to do sentence tokenization with regular expressions, as sentences can end with more than one type of punctuation (for example).

```python
import pandas
import cudf
import re
​
print(s)
0    this is. an? example
dtype: object

s = cudf.Series(['this is. an? example'])
ps = s.to_pandas()
​
print(ps.str.split('[.?!] ', expand=True))
         0   1        2
0  this is  an  example

print(re.split('[.?!] ', ps.iloc[0]))
['this is', 'an', 'example']

# cudf
print(s.str.split('[.?!] '))
                      0
0  this is. an? example

```

### PR DESCRIPTION
**Pull request by davidwendt**
Closes #3584 

This depends on libcudf changes in PR #10128 

This adds the regex parameter to the cudf strings `split()` function similar to the 1.4.0 Pandas one documented [here](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.split.html). 

The main difference is that the `pat` parameter will only be interpreted as regex if the `pat` string has more than 1 character and the `regex` parameter is set to `True`. This is to help with consistency and migration from the previous implementation.

The 1.3.x Pandas version does not have a `regex` parameter for `split()` but instead will try to interpret the intention of the `pat` parameter without it. This seems a bit dangerous since regex would be much slower for us here. Therefore, the `regex` parameter is required to be set to `True` in the cudf implementation in order to use the regex logic path.

Pandas does not support regex for its `rsplit` even though it has been documented and there is an issue here.

### CODE REVIEW

**Comment by davidwendt** [comment]
@shwina I added support for regex in `rsplit` in this PR though it is not supported in Pandas (even in 1.4.0 as far as I can tell). I made it consistent with `split` which accepts a `regex=True` parameter (default to None). Let me know what you think. We can remove the `rsplit` regex support if you think that it would be confusing to have it here.

File: python/cudf/cudf/tests/test_string.py
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_string.py:1548]: Yea, that sounds like a better approach by using `PANDAS_LT_140` from [here](https://github.com/rapidsai/cudf/blob/branch-22.04/python/cudf/cudf/core/_compat.py#L12)

**Review by galipremsagar — APPROVED** [review]

**Comment by davidwendt** [comment]
> I added support for regex in `rsplit` in this PR though it is not supported in Pandas (even in 1.4.0 as far as I can tell). I made it consistent with `split` which accepts a `regex=True` parameter (default to None). Let me know what you think. We can remove the `rsplit` regex support if you think that it would be confusing to have it here.

@shwina Just wanted to check with you on this before merging.

File: python/cudf/cudf/tests/test_string.py
Comment (vyasr) [reply python/cudf/cudf/tests/test_string.py]: This check is actually incorrect right now. While the pandas API documentation for `Series.str.split` includes the regex parameter, as @davidwendt pointed out the function signature doesn't actually support it. However, I just discussed with @shwina offline and we agree that this seems to be an oversight by pandas, especially since `split` does support the regex and the docstring indicates that they _intended_ to support it. We decided that the best path forward is
1. Open a pandas issue on this discrepancy between the API documentation and the actual implementation of `rsplit`.
2. Rewrite this test to manually verify the output rather than comparing to pandas for the cases that pandas does not currently support.

If we want to be fancy we could also make those test cases automatically fail if we detect that the parameters have been added by pandas. In other words, add something like this to the test:
```
# Once this assertion fails, update this test to validate against pandas for sufficiently new pandas versions.
assert 'regex' not in inspect.signature(pd.Series.str.rsplit).parameters.keys()
```

Comment (vyasr) [reply python/cudf/cudf/tests/test_string.py]: CC @davidwendt

Comment (davidwendt) [reply python/cudf/cudf/tests/test_string.py]: For (1): The `rsplit` with regex issue appears to already be an open Pandas issue here: https://github.com/pandas-dev/pandas/issues/29633

Comment (davidwendt) [reply python/cudf/cudf/tests/test_string.py]: For (2): Would it be ok to compare Pandas `rsplit(' ')` to cudf `rsplit('\\s', regex=True)` since these should give equivalent results? 
```
    expect = ps.str.rsplit(pat=" ", n=n, expand=expand)
    got = gs.str.rsplit(pat="\\s", n=n, expand=expand, regex=True)
    assert_eq(expect, got)
```
Or are you looking for something more manual -- checking combinations of `n` and `expand` against a fixed set of Pandas Series or DataFrames?

Comment (vyasr) [reply python/cudf/cudf/tests/test_string.py:1548]: Got it. I think we're fine with (1) then. Regarding (2), if you are testing a particular split that is easily replicated without a regex that sounds fine to me. I thought you might be testing regex splits that can't be replicated with pure string splits, but if not then this seems fine. I don't think you particularly need to test complex regexes since we have plenty of verification on other code paths that the regex code works.

Comment (davidwendt) [reply python/cudf/cudf/tests/test_string.py:1548]: Ok, thanks. I committed the above change to the pytest. Should I wait for another approval or can this be merged?

**Review by vyasr — APPROVED** [review]

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**
